sourcegraph/sourcegraph
# extensions: show extensions decorations in separate columns as experimental feature

First iteration on https://github.com/sourcegraph/sourcegraph/issues/17434

[Design](https://www.figma.com/file/MzN5BUNI0OJ4gQyrKqT7Nw/Improve-git-blame-readability?node-id=246%3A11)

When `experimentalFeatures.enableExtensionsDecorationsColumnView` setting is enabled:
- extension decorations supporting the column view (as for now, git-extras is the only extension supporting the column view: https://github.com/sourcegraph/sourcegraph-git-extras/pull/276) will be rendered in a separate column on the blob page
- extensions decorations are not rendered in the new reference panel blob view
- extension decorations for the diff view and browser extension (native integrations) remain unchanged.

_Context:_ we're investigating the possibility of showing extensions decorations in separate columns ([Figma](https://www.figma.com/file/MzN5BUNI0OJ4gQyrKqT7Nw/Improve-git-blame-readability?node-id=1638%3A3029)). This PR is the first iteration touching only [git-extras extension](https://sourcegraph.com/extensions/sourcegraph/git-extras). We're going to dogfood it internally, then, depending on the feedback, enhance git blame and proceed with other extensions.

Should be merged with https://github.com/sourcegraph/sourcegraph-git-extras/pull/276

|  | Experimental feature disabled | Experimental feature enabled |
| -- | -- | -- |
| Git-extras | ![image](https://user-images.githubusercontent.com/25318659/171332456-8912213b-e294-462e-80e2-5750b699ce22.png) | hover on blame decoration 👇🏻 ![image](https://user-images.githubusercontent.com/25318659/171335128-baffc82d-5cda-4d17-8c85-c6d381d413cb.png) hover on blame decoration when line is selected 👇🏻 ![image](https://user-images.githubusercontent.com/25318659/171335338-5c485caa-f257-4454-9dbf-ea6e049265f5.png) keyboard navigation (Tab, Shift+Tab) when line is selected 👇🏻 ![image](https://user-images.githubusercontent.com/25318659/171334921-f54f3317-e253-4bf6-905e-c31053d4551b.png) |
| Git-extras + codecov | ![image](https://user-images.githubusercontent.com/25318659/171332549-c7dcd451-805d-4635-8281-c1fa4a11dedc.png) | ![image](https://user-images.githubusercontent.com/25318659/171334291-aa2b550b-d05a-4093-8f17-b9a73c827369.png) with tooltip 👇🏻![image](https://user-images.githubusercontent.com/25318659/171334446-92fe2340-28da-4216-b047-884e052e3a66.png) |


## Test plan
- `sg run web-standalone`
- (optional) sideload git-extras extension with https://github.com/sourcegraph/sourcegraph-git-extras/pull/276 revision to have only two blame states: show for the whole file and off (if sideloaded [edit extensionID](https://sourcegraph.com/-/editor?remote_url=git@github.com:sourcegraph/sourcegraph.git&branch=taras-yemets/git-extras-redesign&file=client/web/src/repo/blob/Blob.tsx&editor=VSCode&version=2.2.2&start_row=606&start_col=42&end_row=606&end_col=42&utm_campaign=vscode-extension&utm_medium=direct_traffic&utm_source=vscode-extension&utm_content=vsce-commands) to be `git-extras`)
- ensure git blame on the blob page is displayed in a separate column
- ensure the file diff view and the browser extension blame display remain unchanged



## App preview:

- [Web](https://sg-web-taras-yemets-git-extras-redesign.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-eyfeavboob.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
